src/ripple/shamap/README.md

```diff
@@ -26,164 +45,322 @@ There are two different ways of building and using a SHAMap:
+    == 0)` simply confirms that the TreeNodeCache indeed gave us an immutable
+    node.
 
  2. If the node is not in the TreeNodeCache, we attempt to locate the node
-    in the historic data stored by the data base.  The call to
-    to `fetch(hash)` does that work for us.
+    in the historic data stored by the data base.  The call to to `fetch(hash)`
+    does that work for us.
 
  3. Finally, if ledgerSeq_ is non-zero and we did't locate the node in the
     historic data, then we call a MissingNodeHandler.
 
     The non-zero ledgerSeq_ indicates that the SHAMap is a complete map that
-    belongs to a historic ledger with the given (non-zero) sequence number.
-    So, if all expected data is always present, the MissingNodeHandler should
-    never be executed.
+    belongs to a historic ledger with the given (non-zero, unsharable) sequence
+    number. So, if all expected data is always present, the MissingNodeHandler
+    should never be executed.
 
-    And, since we now know that this SHAMap does not fully represent
-    the data from that ledger, we set the SHAMap's sequence number to zero.
+    And, since we now know that this SHAMap does not fully represent the data
+    from that ledger, we set the SHAMap's sequence number to zero.
 
 If phase 1 returned a node, then we already know that the node is immutable.
 However, if either phase 2 executes successfully, then we need to turn the
 returned node into an immutable node.  That's handled by the call to
-`make_shared<SHAMapTreeNode>` inside the try block.  That code is inside
-a try block because the `fetchNodeExternalNT` method promises not to throw.
-In case the constructor called by `make_shared` throws we don't want to
-break our promise.
+`make_shared<SHAMapTreeNode>` inside the try block.  That code is inside a try
+block because the `fetchNodeExternalNT` method promises not to throw. In case
+the constructor called by `make_shared` throws we don't want to break our
+promise.
 
 
 ## Canonicalize ##
 
-The calls to `canonicalize()` make sure that if the resulting node is already
-in the SHAMap, then we return the node that's already present -- we never
-replace a pre-existing node.  By using `canonicalize()` we manage a thread
-race condition where two different threads might both recognize the lack of a
-SHAMapTreeNode at the same time.  If they both attempt to insert the node
-then `canonicalize` makes sure that the first node in wins and the slower
-thread receives back a pointer to the node inserted by the faster thread.
+`canonicalize()` is called every time a node is introduced into the SHAMap.
+
+A call to `canonicalize()` stores the node in the `TreeNodeCache` if it does not
+already exist in the `TreeNodeCache`.
+
+The calls to `canonicalize()` make sure that if the resulting node is already in
+the SHAMap, node `TreeNodeCache` or database, then we don't create duplicates.
+
+On retrieving existing nodes, the copy in the database has top priority.  If
```

Comment:
I'm a bit confused here, as this seems to be in contradiction with the description of `fetchExternalNodeNT`. In that description, it says we look for the node in the cache first, and then in the database if the node is not in the cache. But here it says the copy in the database has priority over the copy in the cache. Why would we look in the database if the node is already in the cache?

Reply:
The text here was in error, both with what I added, and part of the existing text is obsolete, not describing what `canonicalize` actually does.  Greatly simplified.